s3parq/publish_parq.py

```diff
@@ -166,13 +166,30 @@ def _gen_parquet_to_s3(bucket: str, key: str, dataframe: pd.DataFrame,
         key (str): S3 key to the root of where the dataset should be published
         dataframe (pd.DataFrame): Dataframe that's being published
         partitions (list): List of partition columns
-        
+
     Returns:
         None
     """
     logger.info("Writing to S3...")
-    table = pa.Table.from_pandas(
-        df=dataframe, schema=_parquet_schema(dataframe), preserve_index=False)
+
+    try:
+        table = pa.Table.from_pandas(
+            df=dataframe, schema=_parquet_schema(dataframe), preserve_index=False)
+    except pa.lib.ArrowTypeError:
+        logger.warn(
+            "Dataframe conversion to pyarrow table failed, checking object columns for mixed types")
+        dataframe_dtypes = dataframe.dtypes.to_dict()
+        object_columns = [
+            col for col, col_type in dataframe_dtypes.items() if col_type == "object"]
+        for object_col in object_columns:
+            if not dataframe[object_col].apply(isinstance, args=[str]).all():
+                logger.warn(
+                    f"Dataframe column : {object_col} : in this chunk is type object but contains non-strings, converting to all-string column")
+                dataframe[object_col] = dataframe[object_col].astype(str)
```

Comment:
would it be more efficient to skip the step where you check if everything is a `str` and go straight here to make everything a `str` from the outset?  You have to scan the whole Series with both steps.  I suppose it's just a matter of whether applying `astype(str)` is more costly than `isinstance(str)`

Reply:
So I do think it'd be more efficient but I'm worried about what it'd do to any debugging / hiding issues - if code accidentally put a "" instead of NaN into a float column it'd get silently converted during publish etc.. I wasn't sure of a good way to do so however

Reply:
Ah, yes my solution would hide which columns had multi-type columns.  i'll approve now